models/flowers_oop.nim

```diff
@@ -0,0 +1,473 @@
+  endShaderMode()
+
+  self.a += rand(1.0) * 0.02 - 0.1f
+  if self.a > MaxFlowerParam:
+    self.a = MaxFlowerParam
+  if self.a < MinFlowerParam:
+    self.a = MinFlowerParam
+
+  self.b += rand(1.0) * 0.02 - 0.1f
+  if self.b > MaxFlowerParam:
+    self.b = MaxFlowerParam
+  if self.b < MinFlowerParam:
+    self.b = MinFlowerParam
+
+  self.c += rand(1.0) * 0.5 - 0.25
+  if self.c > 5:
+    self.c = 5
+  if self.c < -5:
+    self.c = -5
+
+  self.d += rand(1.0) * 0.5 - 0.25
+  if self.d > 5:
+    self.d = 5
+  if self.d < -5:
+    self.d = -5
+
+  self.translation = Vector3(
+    x:self.translation.x + self.velocity.x,
+    y:self.translation.y + self.velocity.y,
+    z:self.translation.z + self.velocity.z)
+
+  for f in 0..<3:
+    if system.abs(self.translation.x) > 10:
+      initialize(self)
+    elif system.abs(self.translation.y) > 10:
+      initialize(self)
+    elif system.abs(self.translation.z) > 10:
+      initialize(self)
+    else:
+      self.velocity.x += rand(1.0) * 0.02 - 0.01
+      self.velocity.y += rand(1.0) * 0.02 - 0.01
+      self.velocity.z += rand(1.0) * 0.02 - 0.01
+
+const
+  OrbitSpeed = 0.5f
+  MinDistance = 5.0f
+  MaxDistance = 15.0f
+
+type
+  OrbitCamera = object
+    camera: Camera3D
+    angle: float32
+    distance: float32
+    target: Vector3
+
+proc initOrbitCamera(): OrbitCamera =
+  result = OrbitCamera(
+    camera: Camera3D(
+      position: Vector3(x: 2.0f, y: 2.0f, z: 2.0f),
+      target: Vector3(x: 0.0f, y: 0.0f, z: 0.0f),
+      up: Vector3(x: 0.0f, y: 1.0f, z: 0.0f),
+      fovy: 45.0f,
+      projection: Perspective
+    ),
+    angle: 0.0f,
+    distance: 10.0f,
+    target: Vector3(x: 0.0f, y: 0.0f, z: 0.0f)
+  )
+
+proc updateOrbitCamera(orbit: var OrbitCamera) =
+  # Handle mouse input for rotation
+  if isMouseButtonDown(MouseButton.Left):
+    let deltaX = getMouseDelta().x
+    orbit.angle -= deltaX * OrbitSpeed * getFrameTime()
+
+  # Handle mouse wheel for zoom
+  let wheel = getMouseWheelMove()
+  if wheel != 0.0f:
+    orbit.distance = clamp(orbit.distance - wheel, MinDistance, MaxDistance)
+
+  # Calculate new camera position
+  orbit.camera.position.x = orbit.target.x + sin(orbit.angle) * orbit.distance
+  orbit.camera.position.z = orbit.target.z + cos(orbit.angle) * orbit.distance
+  orbit.camera.position.y = orbit.target.y + orbit.distance * 0.5f
+
+proc main() =
+  # Initialization
+  # --------------------------------------------------------------------------------------
+  initWindow(screenWidth, screenHeight, "JSONverse shaders example - rhodonea")
+  defer: closeWindow()
+
+  var cube = genMeshCube(4.0f, 4.0f, 4.0f)
+  var skybox = loadModelFromMesh(cube)
+  var skyboxShader = loadShaderFromMemory(vertexSkyboxShader, fragmentSkyboxShader)
+  skybox.materials[0].shader = skyboxShader
+
+  var image = loadImage("resources/skybox.png")  # TODO Replace with your texture path
+  if image.data == nil:
+    raise newException(IOError, "Failed to load skybox.png texture")
```

Comment:
Btw these functions have error checking, so you don't need to do that explicitly. Apparently only loadShaderFromMemory is missing a isShaderValid call. Will investigate why.

Reply:
I wouldn't put the checking in myself (I'm kind of bad that way), so I'm glad the library handles it!

Reply:
I wouldn't put the checking in myself (I'm kind of bad that way), so I'm glad the library handles it!